randovania/gui/lib/area_picker.py

```diff
@@ -0,0 +1,262 @@
+from randovania.game_description.world.world_list import WorldList
+from randovania.games.game import RandovaniaGame
+from randovania.game_description.world.node import DockNode, EventNode, GenericNode, Node, NodeIdentifier, PickupNode, ResourceNode, TeleporterNode
+from randovania.gui.generated.area_picker_dialog_ui import Ui_AreaPickerDialog
+from PySide6 import QtWidgets, QtGui, QtCore
+from PySide6.QtCore import Qt
+
+def get_node(header_text: str, parent: QtWidgets.QWidget, game: RandovaniaGame, title_text: str = None, valid_areas: list[AreaIdentifier] = None) -> Optional[NodeIdentifier]:
+    """
+    Opens a modal dialog for the user to pick a specific node.
+
+    :param header_text: the instructional text at the top of the dialog
+    :param parent: the parent widget that spawns the dialog
+    :param game: the game whose nodes will be used
+    :param valid_areas: limit the areas shown in the dialog to the areas in this list
+    :return: a NodeIdentifier of the selected node, or None if the dialog was canceled
+    """
+    accepted, world, area, node = _get_item(header_text, parent, game, valid_areas, True, title_text if title_text else "Select Node")
+    return None if not accepted else NodeIdentifier(AreaIdentifier(world.name, area.name), node.name)
+
+def get_area(header_text: str, parent: QtWidgets.QWidget, game: RandovaniaGame, title_text: str = None, valid_areas: list[AreaIdentifier] = None) -> Optional[NodeIdentifier]:
+    """
+    Opens a modal dialog for the user to pick a specific area.
+
+    :param header_text: the instructional text at the top of the dialog
+    :param parent: the parent widget that spawns the dialog
+    :param game: the game whose areas will be used
+    :param valid_areas: limit the areas shown in the dialog to the areas in this list
+    :return: an AreaIdentifier of the selected area, or None if the dialog was canceled
+    """
+    accepted, world, area, _ = _get_item(header_text, parent, game, valid_areas, False, title_text if title_text else "Select Area")
+    return None if not accepted else AreaIdentifier(world.name, area.name)
+
+def _get_item(header_text: str, parent: QtWidgets.QWidget, game: RandovaniaGame, valid_areas: list[AreaIdentifier] = None, pick_node = True, window_title: str = None) -> tuple[bool, Optional[World], Optional[Area], Optional[Node]]:
+    dialog = AreaPickerDialog(header_text, parent, game, valid_areas, pick_node, window_title)
+    result = dialog.exec()
+    if result == QtWidgets.QDialog.DialogCode.Accepted:
+        return (True, dialog.current_world, dialog.current_area, dialog.current_node)
+    else:
+        return (False, None, None, None)
+
+
+class AreaPickerModel(QtCore.QSortFilterProxyModel):
+    _world_list: WorldList
+    _show_locations: bool = True
+    _show_pickups: bool = True
+    _show_events: bool = True
+    _valid_areas: list[AreaIdentifier]
+    _LOCATION_NODES: list[type] = [DockNode, GenericNode, TeleporterNode]
```

Comment:
Those could be module fields and not class members. And tuples to be immutable too.